File: Design and Analysis of Algorithm/Week1/prog2.py
# 2. Perform binary search on the array: {12, 13, 18, 20, 23} with key = 14.


def binary_search(arr,key):
    low = 0
    high = len(arr)-1
    while low <= high:
        mid = (low+high)//2
        if key < arr[mid]:
            high = mid -1
        elif key > arr[mid]:
            low = mid+1
        else:
            return mid
    return -1


arr = [12, 13, 18, 20, 23]
key = 15

res = binary_search(arr,key)

if(res != -1):
    print("The element is found")
else:
    print("The element is not found")




# Algorithm:

# FUNCTION binary_search(Arr,n,target)
# BEGIN
#     low <- 0                          
#     high <- n-1
#     WHILE low <= high DO
#         mid <- (low+high)//2
#         IF key < arr[mid] THEN
#             high = mid -1
#         ELSE IF key > arr[mid] THEN
#             low = mid+1
#         ELSE
#             RETURN mid
#     END WHILE
#     RETURN -1
# END




# Time Complexity:
# n → n/2 → n/4 → n/8 → … → 1
# T(n) = n/(2^k)
# Now,  n/(2^k) = 1
# =>  log n = k log 2
# =>  k = log n (base 2)
  
# T(n) ~= O(log n)

File: Design and Analysis of Algorithm/Week2/prog2.c
#include <stdio.h>
//Merge sort

void merge(int arr[], int low, int mid, int high){
    int i = low;
    int j = mid+1;
    int k = 0;

    int temp [high-low+1];

    while(i<= mid && j <= high){
        if (arr[i]<= arr[j])
            temp[k++] = arr[i++];
        else    
            temp[k++] = arr[j++];
    }
    while (i<=mid)
        temp[k++] = arr[i++];

    while (j<= high)
        temp[k++] = arr[j++];
    
    for (int g = 0; g < k; g++)
        arr[low + g] = temp[g];
    
    
}

void mergeSort(int arr[], int low, int high){
    if(low < high){
        int mid = (low+high)/2;
        mergeSort(arr,low,mid);
        mergeSort(arr,mid+1,high);
        merge(arr,low,mid,high);
    }
}

void print_result(int arr[], int n) {
    for (int i = 0; i < n; i++)
        printf("%d ", arr[i]);
    printf("\n");
}

int main() {
    int arr[] = {28, 12, 15, 13, 25};
    int n = sizeof(arr) / sizeof(arr[0]);
     mergeSort(arr, 0, n - 1);   
    print_result(arr, n);       
    return 0;
}


// Algorithm : 


// MERGE_SORT(A)
//     if size(A) ≤ 1 then
//         return A

//     mid ← size(A) / 2
//     left ← MERGE_SORT(A[0 … mid−1])
//     right ← MERGE_SORT(A[mid … end])

//     return MERGE(left, right)


// MERGE(left, right)
//     result ← empty list
//     i ← 0, j ← 0

//     while i < length(left) and j < length(right) do
//         if left[i] ≤ right[j] then
//             add left[i] to result
//             i ← i + 1
//         else
//             add right[j] to result
//             j ← j + 1

//     append remaining elements of left to result
//     append remaining elements of right to result

//     return result





// Time Complexity Analysis:

// T(n) = 2T(n/2) + O(n)
//      = (2^k)* T(n/(2^k)) + kn

//      For T(1),  n = 2^k
//               =>k = log n

//         T(n) = n*T(1) + n*log n
//              = O(n log n)